### PR TITLE
`golangci-lint`: enable `prealloc` linter

### PR DESCRIPTION
**Pull request by timvaillancourt**
## What's this?

Enables the `prealloc` linter in golangci-lint to catch slice declarations that could benefit from preallocation, reducing unnecessary allocations.

## Changes

- **212 clean preallocations** added across 145 Go files using `make([]T, 0, len(x))` style
- **31 `//nolint:prealloc` suppressions** where clean preallocation isn't feasible:
  - 17 `var` declarations with non-trivial append patterns (e.g., `1+len(x)` arithmetic)
  - 7 literal initializations followed by append (e.g., `[]T{x}` then `append`)
  - 4 `make([]T, 0)` in nested loops where capacity would require multiplication
  - 3 `= nil` assignments used for intentional state resets
- **4 path-based exclusions** in `.golangci.yml`:
  - `_test.go$` — test files excluded globally
  - `info_schema.go` — hardcoded schema definitions with variable column counts
  - `evalengine/testcases` — test infrastructure
  - `go/tools` — tooling

## Approach

Only preallocate with clean `len()` expressions on slices/maps. No magic number offsets (`len(x)+N`), no hardcoded capacities, no zero-capacity `make`. Anything that doesn't fit this rule gets `//nolint:prealloc`.

---
_Most of this was written by Claude Code — I provided direction and review._

### CODE REVIEW

File: go/vt/mysqlctl/cephbackupstorage/ceph.go [hunk 1]
```diff
@@ -162,11 +162,12 @@ func (bs *CephBackupStorage) ListBackups(ctx context.Context, dir string) ([]bac
 	bucket := alterBucketName(dir)
 
 	// List prefixes that begin with dir (i.e. list subdirs).
-	var subdirs []string
+	doneCh := make(chan struct{})
 	searchPrefix := objName(dir, "")
+	objects := c.ListObjects(bucket, searchPrefix, false, doneCh)
```
Comment (arthurschreiber) [inline go/vt/mysqlctl/cephbackupstorage/ceph.go:162]: Claude pointed out that `objects` is a slice (and it's right). I think this change is not correct.

File: go/mysql/binlog/binlog_json.go [hunk 2]
```diff
@@ -512,12 +512,15 @@ func binparserString(_ jsonDataType, data []byte, pos int) (node *json.Value, er
 // arrays are stored thus:
 // | type_identifier(one of [2,3]) | elem count | obj size | list of offsets+lengths of values | actual values |
 func binparserArray(typ jsonDataType, data []byte, pos int) (node *json.Value, err error) {
-	var nodes []*json.Value
 	var elem *json.Value
 	var elementCount int
 	large := typ == jsonLargeArray
 	elementCount, pos = readInt(data, pos, large)
 	_, pos = readInt(data, pos, large)
+	nodes := make([]*json.Value, 0, elementCount)
```
Comment (arthurschreiber) [inline go/mysql/binlog/binlog_json.go]: Claude suggested using an early return here:

```
	if elementCount == 0 {
		return json.NewArray(nil), nil
	}
```

File: go/sqltypes/marshal.go [hunk 3]
```diff
@@ -142,13 +142,15 @@ func MarshalResult(v any) (*Result, error) {
...
 
+	visibleFields := reflect.VisibleFields(elemType)
+
 	var (
-		exportedStructFields []reflect.StructField
-		fields               []*querypb.Field
-		rows                 []Row
+		exportedStructFields = make([]reflect.StructField, 0, len(visibleFields))
+		fields               = make([]*querypb.Field, 0, len(visibleFields))
+		rows                 = make([]Row, 0, val.Len())
```
Comment (arthurschreiber) [inline go/sqltypes/marshal.go:150]: Claude pointed out here that the `len(visibleFields)` will most likely over-allocate, because not all fields might be exported. I don't think that's an issue, just potentially wasteful.

**Comment by timvaillancourt** [comment]
This PR is also causing this unit test failure because of a change to `nil`-ness:
```bash
2025-11-17T16:59:18.2318158Z === RUN   TestPlanTestSuite/TestPlan/misc_cases.json/execute_no_param_statement
2025-11-17T16:59:18.2318412Z     plan_test.go:704: misc_cases.json
2025-11-17T16:59:18.2318509Z         Diff:
2025-11-17T16:59:18.2318593Z         {
2025-11-17T16:59:18.2318748Z             "Instructions": {
2025-11-17T16:59:18.2318906Z                 "Inputs": [
2025-11-17T16:59:18.2319063Z                     {
2025-11-17T16:59:18.2319517Z                         "FieldQuery": "select 1 from `user` where 1 != 1",
2025-11-17T16:59:18.2319746Z                         "Keyspace": {
2025-11-17T16:59:18.2320009Z                             "Name": "user",
2025-11-17T16:59:18.2320276Z                             "Sharded": true
2025-11-17T16:59:18.2320437Z                         },
2025-11-17T16:59:18.2320725Z                         "OperatorType": "Route",
2025-11-17T16:59:18.2321061Z                         "Query": "select 1 from `user`",
2025-11-17T16:59:18.2321316Z                         "Variant": "Scatter"
2025-11-17T16:59:18.2321444Z                     }
2025-11-17T16:59:18.2321560Z                 ],
2025-11-17T16:59:18.2321775Z                 "OperatorType": "EXECUTE",
2025-11-17T16:59:18.2322220Z                 "Parameters": [0;33mnull => [][0m
2025-11-17T16:59:18.2322315Z             },
2025-11-17T16:59:18.2322531Z             "Original": "execute prep_no_param",
2025-11-17T16:59:18.2322703Z             "QueryType": "EXECUTE",
2025-11-17T16:59:18.2322841Z             "TablesUsed": [
2025-11-17T16:59:18.2322988Z                 "user.user"
2025-11-17T16:59:18.2323087Z             ],
2025-11-17T16:59:18.2323231Z             "Type": "Complex"
2025-11-17T16:59:18.2323309Z         }
2025-11-17T16:59:18.2323391Z         [{
2025-11-17T16:59:18.2323552Z               "Type": "Complex",
2025-11-17T16:59:18.2323742Z               "QueryType": "EXECUTE",
2025-11-17T16:59:18.2323974Z               "Original": "execute prep_no_param",
2025-11-17T16:59:18.2324132Z               "Instructions": {
2025-11-17T16:59:18.2324422Z                 "OperatorType": "EXECUTE",
2025-11-17T16:59:18.2324611Z                 "Parameters": null,
2025-11-17T16:59:18.2324755Z                 "Inputs": [
2025-11-17T16:59:18.2324872Z                   {
2025-11-17T16:59:18.2325116Z                     "OperatorType": "Route",
2025-11-17T16:59:18.2325336Z                     "Variant": "Scatter",
2025-11-17T16:59:18.2325523Z                     "Keyspace": {
2025-11-17T16:59:18.2325734Z                       "Name": "user",
2025-11-17T16:59:18.2325950Z                       "Sharded": true
2025-11-17T16:59:18.2326139Z                     },
2025-11-17T16:59:18.2326528Z                     "FieldQuery": "select 1 from `user` where 1 != 1",
2025-11-17T16:59:18.2326808Z                     "Query": "select 1 from `user`"
2025-11-17T16:59:18.2326918Z                   }
2025-11-17T16:59:18.2327019Z                 ]
2025-11-17T16:59:18.2327124Z               },
2025-11-17T16:59:18.2327271Z               "TablesUsed": [
2025-11-17T16:59:18.2327416Z                 "user.user"
2025-11-17T16:59:18.2327522Z               ]
2025-11-17T16:59:18.2327615Z             }] 
2025-11-17T16:59:18.2327731Z         [{
2025-11-17T16:59:18.2327974Z           "Type": "Complex",
2025-11-17T16:59:18.2328120Z           "QueryType": "EXECUTE",
2025-11-17T16:59:18.2328318Z           "Original": "execute prep_no_param",
2025-11-17T16:59:18.2328447Z           "Instructions": {
2025-11-17T16:59:18.2328622Z             "OperatorType": "EXECUTE",
2025-11-17T16:59:18.2328839Z             "Parameters": [],
2025-11-17T16:59:18.2328967Z             "Inputs": [
2025-11-17T16:59:18.2329062Z               {
2025-11-17T16:59:18.2329275Z                 "OperatorType": "Route",
2025-11-17T16:59:18.2329459Z                 "Variant": "Scatter",
2025-11-17T16:59:18.2329617Z                 "Keyspace": {
2025-11-17T16:59:18.2329791Z                   "Name": "user",
2025-11-17T16:59:18.2329963Z                   "Sharded": true
2025-11-17T16:59:18.2330075Z                 },
2025-11-17T16:59:18.2330405Z                 "FieldQuery": "select 1 from `user` where 1 != 1",
2025-11-17T16:59:18.2330721Z                 "Query": "select 1 from `user`"
2025-11-17T16:59:18.2330823Z               }
2025-11-17T16:59:18.2330911Z             ]
2025-11-17T16:59:18.2331003Z           },
2025-11-17T16:59:18.2331129Z           "TablesUsed": [
2025-11-17T16:59:18.2331250Z             "user.user"
2025-11-17T16:59:18.2331339Z           ]
2025-11-17T16:59:18.2331415Z         }
2025-11-17T16:59:18.2331494Z         ]
```

File: go/sqltypes/marshal.go (hunk 3 above)
Comment (timvaillancourt) [reply go/sqltypes/marshal.go:150]: Yes, this is expected because we sometimes `continue`. But an upper limit should be more efficient than a capacity of zero

File: go/mysql/binlog/binlog_json.go (hunk 2 above)
Comment (timvaillancourt) [reply go/mysql/binlog/binlog_json.go]: Makes sense! Updated 👍

File: go/vt/mysqlctl/cephbackupstorage/ceph.go (hunk 1 above)
Comment (timvaillancourt) [reply go/vt/mysqlctl/cephbackupstorage/ceph.go:162]: Good point. I think it would be better to use a zero capacity here, because we're not really sure

Comment (timvaillancourt) [reply go/vt/mysqlctl/cephbackupstorage/ceph.go:162]: Updated 👍

File: go/vt/vtgate/planbuilder/operators/delete.go [hunk 4]
```diff
@@ -264,7 +263,7 @@ func createDeleteOperator(ctx *plancontext.PlanningContext, del *sqlparser.Delet
 }
 
 func generateOwnedVindexQuery(del *sqlparser.Delete, table TargetTable, ksidCols []sqlparser.IdentifierCI) *sqlparser.Select {
-	var selExprs []sqlparser.SelectExpr
+	selExprs := make([]sqlparser.SelectExpr, 0)
```
Comment (dbussink) [inline go/vt/vtgate/planbuilder/operators/delete.go]: This doesn't seem right? No `len(ksidCols)`?

File: go/vt/vtgate/planbuilder/operators/sharded_routing.go [hunk 5]
```diff
@@ -486,7 +486,7 @@ func (tr *ShardedRouting) processMultiColumnVindex(
 		return newVindexFound
 	}
 
-	var newOption []*VindexOption
+	newOption := make([]*VindexOption, 0, len(v.Options))
```
Comment (dbussink) [inline go/vt/vtgate/planbuilder/operators/sharded_routing.go:489]: I don't know about this specific case, but in general, is it really correct to allocate it all everywhere? Do we not have cases where it's more the exception an element is added and not the common case? And we very much over-allocate? 

That's a general issue with this linter, it's not very smart and complains about every case, while not every case makes really sense to convert.

Comment (timvaillancourt) [reply go/vt/vtgate/planbuilder/operators/sharded_routing.go:489]: > I don't know about this specific case, but in general, is it really correct to allocate it all everywhere?

@dbussink I haven't validated it myself, but I've consistently read _(and just confirmed with ChatGPG)_ that allocating a slice with a fixed capacity is "usually" more efficient, because a zero-capacity slice has to be grown every append, causing new allocations for a new backing array each append

To my knowledge the risk to using a capacity that may/may not be 100% used is memory usage that may just be filled with empty values. But there must be cases where this isn't always true, perhaps really huge slices(?). I wondered if you had any scenarios in mind?

File: go/vt/vtgate/planbuilder/operators/delete.go (hunk 4 above)
Comment (timvaillancourt) [reply go/vt/vtgate/planbuilder/operators/delete.go]: @dbussink I kept capacity zero here because of this sub-loop that I don't really know the ~~capacity~~ length of up-front: https://github.com/vitessio/vitess/blob/9e5aeb66724488ac334e180f98b4822fe3f17e06/go/vt/vtgate/planbuilder/operators/delete.go#L272-L275

**Comment by timvaillancourt** [comment]
👋 all, I dusted this PR off and ask Claude to help me debug the failures it caused

The approach is now dialed-back a bit: if it's easy for us to know the length to pre-allocate, we do that. By "easy" I mean if there is a known `len(something)` that we can use to know capacity

In cases where we need to manually count our `append`s, the linter is ignored. It's not a very nice developer experience to need to count `append`s, at least for now

Finally, if the code was simply incompatible with zero-length slices vs. `nil`, I chose to ignore the linter violations. The ignored violations are in the config file for us to follow-up on later

Claude has updated the PR summary with the latest state of things 👍

File: go/vt/wrangler/vexec.go [hunk 6]
```diff
@@ -102,10 +102,10 @@ func (wr *Wrangler) QueryResultForRowsAffected(results map[*topo.TabletInfo]*sql
 	}}
 	var row2 []sqltypes.Value
 	for tablet, result := range results {
-		row2 = nil
 		row2 = append(row2, sqltypes.NewVarBinary(tablet.AliasString()))
 		row2 = append(row2, sqltypes.NewUint64(result.RowsAffected))
 		qr.Rows = append(qr.Rows, row2)
+		row2 = row2[:0] // reset the slice
```
Comment (mhamza15) [inline go/vt/wrangler/vexec.go:108]: I believe all rows will now share the same backing array, so each iteration overwrites previous rows.

File: go/vt/vtctl/vdiff2.go [hunk 7]
```diff
@@ -299,18 +299,19 @@ func getStructFieldNames(s any) []string {
...
+	var str string
 	fields := getStructFieldNames(VDiffListing{})
+	strArray := make([]string, 0, len(fields))
+	strArray2 := make([][]string, 0, len(listings))
+
 	strArray = append(strArray, fields...)
 	strArray2 = append(strArray2, strArray)
 	for _, listing := range listings {
-		strArray = nil
+		strArray = strArray[:0]
```
Comment (mhamza15) [inline go/vt/vtctl/vdiff2.go:314]: Same comment as above.

File: go/vt/vtgate/semantics/info_schema.go [hunk 8]
```diff
@@ -310,60 +310,59 @@ func getInfoSchema57() map[string][]vindexes.Column {
...
-	cols = []vindexes.Column{}
+	cols = nil
 	cols = append(cols, createCol(parser, "DOC_ID", 265, "utf8mb3_general_ci", "0", 21, 0, true, ""))
 	infSchema["INNODB_FT_BEING_DELETED"] = cols
-	cols = []vindexes.Column{}
+	cols = nil
 	cols = append(cols, createCol(parser, "KEY", 6165, "utf8mb3_general_ci", "", 193, 0, true, ""))
 	cols = append(cols, createCol(parser, "VALUE", 6165, "utf8mb3_general_ci", "", 193, 0, true, ""))
 	infSchema["INNODB_FT_CONFIG"] = cols
-	cols = []vindexes.Column{}
```
Comment (mhamza15) [inline go/vt/vtgate/semantics/info_schema.go:344]: Missing `cols = nil` instead?

**Comment by timvaillancourt** [comment]
👋 converted this to draft, and I've broken the change into VTOrc-only for now: https://github.com/vitessio/vitess/pull/19665

This PR can act as a "map" for a few smaller PRs to rollout `prealloc` 👍

**Comment by timvaillancourt** [comment]
This is happening in stages in a handful of PRs. Closing
